web/api/rest-api-design/Batching-Filtering-Pagination.md

```diff
@@ -0,0 +1,73 @@
+## Batching, Filtering, and Pagination
+
+### Filtering:
+
+- Topics covered: Elasticsearch and URL Parameters
+
+#### Elasticsearch:
+
+- Elasticsearch is a distributed RESTful search engine that is recommended for most filtering needs.
+- It works by sorting results by a relevance score, which is a way to quantify how well each result matches a query (higher score = more relevance, e.g. the more relevant the document being searched)
+- The score is also determined by whether the query is run via a query or filter context.
+- In a **query context** the following query clause is passed to a query parameter to calculate a relevance score: How well does this document match this query clause?
+- When a **filter context** is used, which happens mostly when filtering structured data, a query clause is passed to a filter parameter (e.g. `filter`). Some examples include:
+
+  - Does this timestamp fall into the range 2015 to 2016?
+  - Is the status field set to "published"?
+
+Examples of a query and filter context:
+
+```
+GET /_search
+{
+  "query": {
+    "bool": {
+      "must": [
+        { "match": { "title":   "Search" }},
+        { "match": { "content": "Elasticsearch" }}
+      ],
+      "filter": [
+        { "term":  { "status": "published" }},
+        { "range": { "publish_date": { "gte": "2015-01-01" }}}
+      ]
+    }
+  }
+}
+
+```
+
+In this example we see the query parameter `”query”:{...}` indicates a context. The `bool` and two `match clauses` within the query context are used to score the document relevance. Finally, the filter parameters `”filter”: [...]` have two clauses, `term` and `range`, which are part of the filter context and allow it to filter out the documents that don’t match.
+
+#### URL parameters:
+
+- You can filter using resource names. For example, if you have an API that tracks different dogs breeds, their names, and their birth years.
+
+`GET https://api.thedogapi.com/v1/breeds`
+
+- If adding a query to your url parameters, use question marks before providing a query parameters, e.g. limits:
+  `GET https://api.thedogapi.com/v1/breeds?limit=2`
+- Square brackets on key names within a resource provide more flexibility with filtering and enable you to indicate a range. For example, one way to indicate a range of values:
```

Comment:
This is interesting. The way I've implemented brackets in the APIs I've designed and the ones I've used, was to specify multiple options as an OR query. There typically isn't anything inside the bracket. For example, if I want to search for repos that match a particular language, I would use a query like this: `https://www.example.com/repos?language[]=ruby&language[]=markdown`. This would show me all repos that use ruby, as well as all repos that use markdown. As opposed to only repos that use BOTH ruby and markdown in the same repo.

I don't think I've seen brackets and the ampersand used to indicate a **range**.